core/src/main/java/io/grpc/internal/AbstractServerStream.java

```diff
@@ -31,256 +31,207 @@
 
 package io.grpc.internal;
 
-import static com.google.common.base.Preconditions.checkNotNull;
-
 import com.google.common.base.Preconditions;
 
 import io.grpc.Attributes;
 import io.grpc.Metadata;
 import io.grpc.Status;
 
-import java.io.InputStream;
-import java.util.logging.Level;
-import java.util.logging.Logger;
+import javax.annotation.Nullable;
 
 /**
- * Abstract base class for {@link ServerStream} implementations.
- *
- * @param <IdT> the type of the stream identifier
+ * Abstract base class for {@link ServerStream} implementations. Extending classes only need to
+ * implement {@link #transportState()} and {@link #abstractServerStreamSink()}.
  */
-public abstract class AbstractServerStream<IdT> extends AbstractStream<IdT>
-    implements ServerStream {
-  private static final Logger log = Logger.getLogger(AbstractServerStream.class.getName());
-
-  /** Whether listener.closed() has been called. */
-  private boolean listenerClosed;
-  private ServerStreamListener listener;
-
-  private boolean headersSent = false;
+public abstract class AbstractServerStream extends AbstractStream2
+    implements ServerStream, MessageFramer.Sink {
   /**
-   * Whether the stream was closed gracefully by the application (vs. a transport-level failure).
+   * Only called from application thread. A separate interface just to avoid name
+   * collisions/confusion.
    */
-  private boolean gracefulClose;
-  /** Saved trailers from close() that need to be sent once the framer has sent all messages. */
-  private Metadata stashedTrailers;
+  protected interface Sink {
+    /**
+     * Sends response headers to the remote end points.
+     *
+     * @param headers the headers to be sent to client.
+     */
+    void writeHeaders(Metadata headers);
 
-  protected AbstractServerStream(WritableBufferAllocator bufferAllocator,
-                                 int maxMessageSize) {
-    super(bufferAllocator, maxMessageSize);
+    /**
+     * Sends an outbound frame to the remote end point.
+     *
+     * @param frame a buffer containing the chunk of data to be sent.
+     * @param flush {@code true} if more data may not be arriving soon
+     */
+    void writeFrame(@Nullable WritableBuffer frame, boolean flush);
+
+    /**
+     * Sends trailers to the remote end point. This call implies end of stream.
+     *
+     * @param trailers metadata to be sent to end point
+     * @param headersSent {@code true} if response headers have already been sent.
```

Comment:
> Why does the application need to keep track of whether the initial headers have been sent?

Firstly, I didn't change any existing API surface as exposed to the application, and it was a design goal not to.

Secondly, the application doesn't need to keep track of it. This method is normally only be called from within the stream, which can be seen a little since `transportState()` has protected visibility. The only caller is intended to be `AbstractServerStream.writeHeaders()`, which tracks `headersSent`.

Keeping track of `headersSent` in `AbstractServerStream` is not strictly necessary, but most server implementations would need that information, so it is done as a convenience.

> Having this extra boolean seems wrong. If the application has to keep track of this anyway, the API might as well be: writeHeaders(Metadata metadata, boolean isTrailers);

Yes and no. "Headers" and "Trailers" have special meaning in gRPC, in that Status always arrives on Trailers, and that you can receive only Trailers. The boolean is there to help mapping to HTTP which has slightly different semantics, but not completely different semantics (in that Trailers in gRPC will always have endStream=true in HTTP/2). As I said above, it's a convenience and it isn't exposed to the application.

Some of this is also left over from when ClientCallImpl did not enforce proper call order, which was instead done by `AbstractStream`/`AbstractServerStream`. If the application sent trailers and then messages it would have needed to result in an exception. So maybe it could be changed now, but I still wouldn't want to change it in this PR.